src/freertos_drivers/esp32/Esp32WiFiManager.cxx

```diff
@@ -0,0 +1,985 @@
+    ai->ai_flags = 0;
+    ai->ai_family = hints->ai_family;
+    ai->ai_socktype = hints->ai_socktype;
+    ai->ai_protocol = hints->ai_protocol;
+    ai->ai_addrlen = sizeof(struct sockaddr_in);
+    sa_in->sin_len = sizeof(struct sockaddr_in);
+    sa_in->sin_family = hints->ai_family;
+
+    string service_name = service, protocol_name;
+    SPLIT_MDNS_SERVICE_NAME(service_name, protocol_name)
+
+    mdns_result_t *results = NULL;
+    esp_err_t err = mdns_query_ptr(service_name.c_str(), protocol_name.c_str(),
+        MDNS_QUERY_TIMEOUT, MDNS_MAX_RESULTS, &results);
+    LOG(VERBOSE, "[mDNS] mdns_query_ptr: %s.", esp_err_to_name(err));
+    if(err)
+    {
+        // failed to find any matches
+        LOG(WARNING, "[mDNS] mDNS query failed: %s.", esp_err_to_name(err));
+        return EAI_FAIL;
+    }
+
+    if(!results)
+    {
+        // failed to find any matches
+        LOG(WARNING, "[mDNS] No matches found for service: %s.", service);
+        return EAI_AGAIN;
+    }
+
+    // make a copy of the results to preserve the original list for cleanup.
+    mdns_result_t *res = results;
+    // scan the mdns query results linked list, the first match with an IPv4
+    // address will be returned.
+    bool match_found = false;
+    while(res && !match_found)
+    {
+        mdns_ip_addr_t *ipaddr = res->addr;
+        while(ipaddr && !match_found)
+        {
+            // if this result has an IPv4 address process it
+            if(ipaddr->addr.type == IPADDR_TYPE_V4)
+            {
+                LOG(INFO, "[mDNS] Found %s as providing service: %s on port %d.",
+                    res->hostname, service, res->port);
+                inet_addr_from_ip4addr(&sa_in->sin_addr,
+                    &ipaddr->addr.u_addr.ip4);
+                sa_in->sin_port = htons(res->port);
+                match_found = true;
+            }
+            ipaddr = ipaddr->next;
+        }
+        res = res->next;
+    }
+
+    // free up the query results linked list.
+    mdns_query_results_free(results);
+
+    if(!match_found)
+    {
+        LOG(WARNING, "[mDNS] No matches found for service: %s.", service);
+        return EAI_AGAIN;
+    }
+
+    // return the resolved data to the caller
+    *addr = ai.release();
+    (*addr)->ai_addr = sa.release();
+
+    // successfully resolved an address, inform the caller
+    return 0;
+}
+
+// The functions below are not available via the standard ESP-IDF provided
+// API.
+
+/// Retrieves the IPv4 address from the ESP32 station interface.
+///
+/// @param ifap will hold the IPv4 address for the ESP32 station interface when
+/// successfully retrieved.
+/// @return zero for success, -1 for failure.
+int getifaddrs(struct ifaddrs **ifap)
+{
+    tcpip_adapter_ip_info_t ip_info;
+
+    /* start with something "safe" in case we bail out early */
+    *ifap = nullptr;
+
+    if(!tcpip_adapter_is_netif_up(TCPIP_ADAPTER_IF_STA))
+    {
+        // Station TCP/IP interface is not up
```

Comment:
what about 127.0.0.1, should that be included in the interfaces?

Reply:
I'm not certain that the esp32 even has an equivalent to localhost. From the tcpadapter code there is only STA, AP and ETH interfaces. I don't should support ETH in this code, I'm also uncertain what components would be needed to support that if we decided to via a more generic interface (NetworkInterface, WiFiInterface extending NetworkingInterface, Esp32WiFi extending WiFiInterface, EspEthernet extending NetworkInterface, etc)